### PR TITLE
Remove ClassCallCheck, possibleConstructorReturn in loose mode

### PR DESCRIPTION
**Pull request by hzoo**
| Q                                   | A  
| -----------------------| ---
| Bug fix?                        | 
| Breaking change?        | 
| New feature?                | yes
| Deprecations?              | 
| Spec compliancy?        | 
| Tests added + pass?    | yes
| Fixed tickets                 |
| License                         | MIT
| Doc PR                          | 
| Dependency Changes | 

cc @developit

### CODE REVIEW

**Comment by loganfsmyth** [comment]
Wasn't it `possibleConstructorReturn` that @developit was asking about?

**Comment by developit** [comment]
Both actually. I'm looking for a way to say "hey, don't worry about protecting classes, they are internal to my module anyway".

**Comment by DrewML** [comment]
Not that I have anything against this change, but I'm just curious to know what the use-case is @developit. Performance?

**Comment by developit** [comment]
Performance and file size. I use codemods or avoid classes currently, with this in place it would be more viable.

**Comment by hzoo** [comment]
@developit oh and you can also check out babili's plugins + babel codemods

**Comment by DrewML** [comment]
Part of me wonders if these things should just be part of `loose` mode for `es2015-classes`, rather than separate options. Thoughts @hzoo?

**Comment by developit** [comment]
@hzoo think babili would cover stripping these? That would get me to upgrade to 6 for sure! 😂

**Comment by developit** [comment]
@DrewML huge fringe benefit of doing that would be that the `es2015` preset would be able to enable these mods with a single option. Currently disassembling that preset is a bit of a hard battle see modify-babel-preset haha). 

Perhaps another option is a `"minimal"` mode? Or maybe that's encroaching on Babili territory.

**Comment by DrewML** [comment]
I think `minimal` mode was the initial intent of `loose` (but who knows at this point). 

Babel 5 docs on loose mode were:

> Loose mode enables certain transformers to generate cleaner output that lacks specific ES6 edgecases. These edgecases are either unlikely to appear in your code or the inclusion of them introduces significant overhead.

> As a result of loose mode code will execute faster and be much more readable and comparable to the original but will deviate from the spec in slight ways.

I kind of interpret that as "generate the minimal amount of changes to make the code work, and ignore edge-cases where someone could do something bad"

**Comment by developit** [comment]
That is certainly how I think of it. It would be a very compelling option to have handy, at least for me. Tends to be like 300-500 bytes, which for small libs like [preact](https://preactjs.com) is a heck of a lot (10%!).

**Comment by DrewML** [comment]
@hzoo Thoughts/Opinions on combining this into loose mode?

**Comment by hzoo** [comment]
We could? I guess it would be a "breaking change" still? although most that use loose mode wouldn't care about it anyway

**Comment by developit** [comment]
It be breaking for anything that is relying on calling constructors to throw (for whatever reason, I can't think of a good one, maybe some over-testing?). It's just the preset though and I would guess as you said, those are not the people running in loose mode. Also since loose was only added in 6.13 (?) it seems like it'd be not quite as widely used yet.

**Comment by hzoo** [comment]
This changes the plugin not the preset so it would affect everyone using the option which should of been there since the beginning? Yeah it's removing a throw check so it wouldn't break anything in the opposite way

**Comment by DrewML** [comment]
I guess this kind of leads to the bigger question of "what is a breaking change" in these options. In theory, we could _never_ add any features to `loose` mode, because any additional optimization could always be a breaking change.

Semver is hard :(

**Comment by Download** [comment]
Using `loose` mode is the way to go for both `classCallCheck` and `possibleConstructorReturn` if you ask me. 

I would like for Babel to generate code that is as close to ES6 compliant as possible under normal circumstances, then use loose mode to drop every line of code that is not strictly necessary to make stuff just work. It means in loose mode you can do stuff that is against spec... But probably you will dev in strict mode and catch such uses there. Once it's in production, use loose mode for max. performance.

Having a single toggle on the preset would be infinitely more valuable than having to control each plugin separately.

**Comment by developit** [comment]
Also makes fall-through from presets like `latest` easier since it's just the existing `loose` option, right?

**Comment by hzoo** [comment]
Yes it should although fyi `latest` is deprecated since `env` covers all it's functionality

**Comment by developit** [comment]
You know I'm behind on all that stuff (still using 5 in places haha)

**Comment by hzoo** [comment]
😅 can someone just make the auto upgrade from 5 to 6 bot

**Comment by developit** [comment]
@hzoo still produces larger output for some reason, need to look into why. I'm basically only using babel to convert `let` and `const` to var without caring about TDZ or block scoping.

File: packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js]: Do we want to remove the checks here too?

File: packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+let B = function B() {};
+
+let C = function (_B) {
+  _inherits(C, _B);
+
+  function C() {
+    woops.super.test();
+
+    var _this = _B.call(this) || this;
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js]: can drop `|| this` as well

File: packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js (hunk 1 above)
Comment (developit) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js]: Definitely. My view here is that people using loose mode classes are just asking for a direct replacement for the ES5 equivalent.

**Comment by Download** [comment]
@hzoo Loving what you are doing here! Can't wait for it to be merged and released!

File: packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
+
+let B = function B() {};
+
+let A = function (_B) {
+  _inherits(A, _B);
+
+  function A(track) {
+    if (track !== undefined) {
+      var _this = _B.call(this, track) || this;
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose-classCallCheck/with-superClass/expected.js]: and remove `|| this` I guess?

File: packages/babel-helpers/src/helpers.js [hunk 4]
```diff
@@ -377,6 +377,16 @@ helpers.inherits = template(`
   })
 `);
 
+helpers.inheritsLoose = template(`
+  (function (subClass, superClass) {
+    subClass.prototype = Object.create(superClass && superClass.prototype);
+    subClass.prototype.constructor = subClass;
+    if (superClass) {
```
Comment (developit) [inline packages/babel-helpers/src/helpers.js]: Thoughts on also removing this check?

Comment (hzoo) [inline packages/babel-helpers/src/helpers.js]: because we would assume superClass is not null?

Comment (developit) [inline packages/babel-helpers/src/helpers.js]: good call.

File: packages/babel-helpers/src/helpers.js [hunk 5]
```diff
@@ -377,6 +377,16 @@ helpers.inherits = template(`
   })
 `);
 
+helpers.inheritsLoose = template(`
+  (function (subClass, superClass) {
+    subClass.prototype = Object.create(superClass && superClass.prototype);
```
Comment (hzoo) [inline packages/babel-helpers/src/helpers.js]: I guess we would be able to remove the check here too then @developit?

File: packages/babel-plugin-transform-es2015-classes/test/fixtures/loose/accessing-super-class/expected.js [hunk 6]
```diff
@@ -1,20 +1,18 @@
...
 
   function Test() {
     var _Foo$prototype$test, _Foo$prototype$test2;
 
-    babelHelpers.classCallCheck(this, Test);
-
     woops.super.test();
 
-    var _this = babelHelpers.possibleConstructorReturn(this, _Foo.call(this));
+    var _this = _Foo.call(this) || this;
```
Comment (hzoo) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose/accessing-super-class/expected.js]: should we remove ` || this`?

Comment (jridgewell) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose/accessing-super-class/expected.js]: If we do, we'd have to remove the whole `var _this = ` (and all the `_this`s) too. I think this is fine for now.

File: packages/babel-helpers/src/helpers.js (hunk 4 above)
Comment (hzoo) [inline packages/babel-helpers/src/helpers.js]: Ok we can just assume in loose mode that the superClass is non-null and has a prototype to get

```js
  (function (subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    subClass.__proto__ = superClass;
  })
```

**Comment by hzoo** [comment]
Ok will merge and someone can do https://github.com/babel/babel/pull/4850#discussion_r123809671 as a followup issue/PR

**Comment by hzoo** [comment]
Also something about travis is messed up (circle ci passes + local)

File: packages/babel-helpers/src/helpers.js (hunk 5 above)
Comment (developit) [inline packages/babel-helpers/src/helpers.js]: yup

File: packages/babel-plugin-transform-es2015-classes/test/fixtures/loose/accessing-super-class/expected.js (hunk 6 above)
Comment (developit) [inline packages/babel-plugin-transform-es2015-classes/test/fixtures/loose/accessing-super-class/expected.js]: hmm. this seems important.

**Comment by Download** [comment]
@hzoo Is there a way to test this in my project before it is released to NPM?

**Comment by hzoo** [comment]
I could some better docs for this (or a utility to make it easy) but yeah you can just overwrite your node_modules? Like literally replace the 2 files that we changed. You can probably just run the src/index.js through babeljs.io/repl and then replace the lib/index.js in node_modules and add that helper to test it

**Comment by Download** [comment]
@hzoo Ok. I guess that's a bit too far outside of my comfort zone :) I guess I'll wait until the next alpha makes it to NPM.  :)

**Comment by hzoo** [comment]
https://github.com/babel/babel/releases/tag/v7.0.0-alpha.15

**Pull request merged**